moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h

```diff
@@ -217,6 +220,185 @@ struct CollisionRequest
+    , active_components_only(active_components_only)
+    , acm(acm)
+    , distance_threshold(distance_threshold)
+    , verbose(false)
+    , gradient(false)
+  {
+  }
+  DistanceRequest(bool detailed, bool global, const std::set<const robot_model::LinkModel*>& active_components_only,
+                  const AllowedCollisionMatrix& acm, double distance_threshold = std::numeric_limits<double>::max())
+    : detailed(detailed)
+    , global(global)
+    , active_components_only(&active_components_only)
+    , acm(&acm)
+    , distance_threshold(distance_threshold)
+    , verbose(false)
+    , gradient(false)
+  {
+  }
+  DistanceRequest(bool detailed, bool global, const std::string group_name, const AllowedCollisionMatrix* acm,
+                  double distance_threshold = std::numeric_limits<double>::max())
+    : detailed(detailed)
+    , global(global)
+    , group_name(group_name)
+    , active_components_only(NULL)
+    , acm(acm)
+    , distance_threshold(distance_threshold)
+    , verbose(false)
+    , gradient(false)
+  {
+  }
+  DistanceRequest(bool detailed, bool global, const std::string group_name, const AllowedCollisionMatrix& acm,
+                  double distance_threshold = std::numeric_limits<double>::max())
+    : detailed(detailed)
+    , global(global)
+    , group_name(group_name)
+    , active_components_only(NULL)
+    , acm(&acm)
+    , distance_threshold(distance_threshold)
+    , verbose(false)
+    , gradient(false)
+  {
+  }
+
+  virtual ~DistanceRequest()
+  {
+  }
+
+  /// Compute \e active_components_only_ based on \e req_
+  void enableGroup(const robot_model::RobotModelConstPtr& kmodel)
+  {
+    if (kmodel->hasJointModelGroup(group_name))
+      active_components_only = &kmodel->getJointModelGroup(group_name)->getUpdatedLinkModelsWithGeometrySet();
+    else
+      active_components_only = NULL;
+  }
+
+  /// Indicate if detailed information should be calculated
+  bool detailed;
+
+  /// Indicate if the global minimum should be found
+  bool global;
+
+  std::string group_name;
+
+  /// The set of active components to check
+  const std::set<const robot_model::LinkModel*>* active_components_only;
+
+  /// The allowed collision matrix used to filter checks
+  const AllowedCollisionMatrix* acm;
+
+  /// A distance threshold to reduce number of queries
+  double distance_threshold;
+
+  /// Log debug information
+  bool verbose;
+
+  /// Indicate if gradient should be calculated between each object. This the vector defined by the line connecting the
+  /// closest points on the two objects.
+  bool gradient;
+};
+
+struct DistanceResultsData
+{
+  DistanceResultsData()
+  {
+    clear();
+  }
+
+  /// The minimum distance between two objects. If two objects are in collision, min_distance <= 0.
+  double min_distance;
+
+  /// The nearest points
+  Eigen::Vector3d nearest_points[2];
+
+  /// The object link names
+  std::string link_name[2];
+
+  /// The gradient
+  Eigen::Vector3d gradient;
```

Comment:
Specify the direction of the gradient: `nearest_point[1] - nearest_point[0]` or the other way around?
Is the vector normalized?